src/auto-ssh-agent/pam_auto_ssh_agent.c

```diff
@@ -0,0 +1,756 @@
+      setegid (pwd->pw_gid) < 0 || seteuid (pwd->pw_uid) < 0)
+    {
+      error ("couldn't setup credentials: %s",
+              strerror (errno));
+      exit (EXIT_FAILURE);
+    }
+
+  /* Setup environment variables */
+  ret = setup_pam_env (pamh, "HOME", pwd->pw_dir);
+  if (ret == PAM_SUCCESS)
+    {
+      ret = setup_pam_env (pamh, "PATH", "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin");
+    }
+  if (ret == PAM_SUCCESS && !pam_getenv (pamh, "DISPLAY"))
+    {
+      display = getenv ("DISPLAY");
+      if (display)
+        ret = setup_pam_env (pamh, "DISPLAY", display);
+    }
+  if (ret == PAM_SUCCESS && !pam_getenv (pamh, "XDG_RUNTIME_DIR"))
+    {
+      runtime = getenv ("XDG_RUNTIME_DIR");
+      if (runtime)
+        ret = setup_pam_env (pamh, "XDG_RUNTIME_DIR", runtime);
+    }
+
+  /* Make sure that worked */
+  if (ret != PAM_SUCCESS)
+    {
+      error ("couldn't setup environment: %s",
+              pam_strerror (pamh, ret));
+       exit (EXIT_FAILURE);
+    }
+
+  /* Now actually execute the process */
+  execve (args[0], args, pam_getenvlist (pamh));
+  error ("couldn't run %s: %s", args[0], strerror (errno));
+  exit (EXIT_FAILURE);
+}
+
+static void
+ignore_signals (struct sigaction *defsact,
+                struct sigaction *oldsact,
+                struct sigaction *ignpipe,
+                struct sigaction *oldpipe)
+{
+  /*
+   * Make sure that SIGCHLD occurs. Otherwise our waitpid below
+   * doesn't work properly. We need to wait on the process to
+   * get the daemon exit status.
+   */
+  memset (defsact, 0, sizeof (*defsact));
+  memset (oldsact, 0, sizeof (*oldsact));
+  defsact->sa_handler = SIG_DFL;
+  sigaction (SIGCHLD, defsact, oldsact);
+
+  /*
+   * Make sure we don't exit with a SIGPIPE while doing this, that
+   * would be very annoying to a user trying to log in.
+   */
+  memset (ignpipe, 0, sizeof (*ignpipe));
+  memset (oldpipe, 0, sizeof (*oldpipe));
+  ignpipe->sa_handler = SIG_IGN;
+  sigaction (SIGPIPE, ignpipe, oldpipe);
+}
+
+static void
+restore_signals (struct sigaction *oldsact,
+                 struct sigaction *oldpipe)
+{
+  /* Restore old handler */
+  sigaction (SIGCHLD, oldsact, NULL);
+  sigaction (SIGPIPE, oldpipe, NULL);
+}
+
+static pid_t
+run_as_user (pam_handle_t *pamh,
+             struct passwd *pwd,
+             char **args,
+             int inp[2],
+             int outp[2],
+             int errp[2])
+{
+  pid_t pid = -1;
+
+  /* Start up daemon child process */
+  switch (pid = fork ())
+    {
+    case -1:
+      error ("couldn't fork: %s", strerror (errno));
+      goto done;
+
+    /* This is the child */
+    case 0:
+      setup_child (inp, outp, errp, pamh, pwd, args);
+      /* Should never be reached */
+      break;
+
+    /* This is the parent */
+    default:
+      break;
+    };
+
+done:
+  return pid;
+}
+
+static int
+setup_environment_from_agent (char *line,
+                              void *arg)
```

Comment:
We still would need the pid of the process in order to kill it at the end of the session. Because it forks, the pid it launches with isn't the same as the eventual agent pid. This seemed like the best way to get it.